### ISSUE TITLE
Example code gives: TypeError: Object(...) is not a function

### ISSUE THREAD

**Issue by hindsc52** [jackyef/react-isomorphic-data#48]
Hi, Love this idea behind this repo.

I tried out the ssr example code you have in the readme and I get this error from the try catch around the getTreeData method.

`TypeError: Object(...) is not a function`

**Comment by jackyef** [comment]
Hi @hindsc52, glad that the project is helping someone!

I am not sure why you closed the issue, did you resolve the issue already? If you haven't, would you me out by providing a reproduction for the case?

Thanks!

**Comment by hindsc52** [comment]
Hi @jackyef apologies for opening and closing an issue so quickly. I was following the SSR example on your main readme, but I missed the use case of the using `renderToStingWithData()`

I found this library as I am looking for a declarative way to handle data fetching from components whilst support SSR and client side hydration.

I currently fetch the main page data and pass that in as `initialProps` however some components require data from other apis, and it would nice to be able to fetch all of this data at once and then send it to the react app to ssr.

Reacts support for SSR still seems fairly half baked right now and data fetching when it comes to full server side renders is actually fairly difficult

**Comment by jackyef** [comment]
No problem. Could you provide the code you are trying out here so I could take a look?

The function name should be `renderToStringWithData` not `renderToStingWithData`.

**Comment by hindsc52** [comment]
> No problem. Could you provide the code you are trying out here so I could take a look?
> 
> The function name should be `renderToStringWithData` not `renderToStingWithData`.

Oops spelling mistake on my side :)

The code I tied was this from your docs site examples

```
import express from 'express';
import React from 'react';
import { renderToString } from 'react-dom/server';
import { DataProvider, createDataClient } from 'react-isomorphic-data';
import { getDataFromTree } from 'react-isomorphic-data/ssr';
import fetch from 'node-fetch';
import App from './App';
// react-isomorphic-data needs fetch to be available in the global scope
global.fetch = fetch;
express.get('/*', async (req: express.Request, res: express.Response) => {
  const dataClient = createDataClient({ 
    initialCache: {}, 
    ssr: true, 
    headers: {
      // forward headers from client to the REST API (such as cookies)
      'cookie': req.header('cookie'),
      'my-custom-header': 'will be sent on all requests',
  }});
  const reactApp = (
    <DataProvider client={dataClient}>
      <App />
    </DataProvider>
  );
  try {
    await getDataFromTree(reactApp, dataClient);
  } catch (err) {
    console.error('Error while trying to getDataFromTree', err);
  }
  const markup = renderToString(reactApp);
  res.send(
    `<!doctype html>
      <html lang="">
      <head>
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta charSet='utf-8' />
        <title>Razzle TypeScript</title>
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <script>
          window.__cache=${JSON.stringify(dataClient.cache)}
        </script>
      </head>
      <body>
        <div id="root">${markup}</div>
      </body>
    </html>`
  );
}
```

This produced the error detailed in this issue. However the SSR example from this repo worked just as expected.

I going to try and implement this with a codebase I am working that has a much more complex SSR setup.

Am i right in thinking that ultimately this package will look at each component i the react app tree and see which ones use the useData hook or the withData HOC and fetches the data from the specified endpoints then saving the data in a context store?

**Comment by jackyef** [comment]
Ah, did you copy that snippet to the examples in this repo?
There are some incorrect code there, I will have to fix that.

But, the snippet is just a simple implementation, but it's not supposed to work just by copy pasting it. `import App from './App` might not be there if you are just copying it to any project. I guess I should have made a snippet that work directly when copied and pasted?

>Am i right in thinking that ultimately this package will look at each component i the react app tree and see which ones use the useData hook or the withData HOC and fetches the data from the specified endpoints then saving the data in a context store?
Kinda. To do that in React, we are actually calling `renderToStaticMarkup` repeatedly to know which data we need to fetch. So it goes like:
1. Render the app, collect all the fetch promises along the way
2. Resolve all the promises, add the data to cache
3. Render the app again. This time, we might render deeper into the component tree, and might find new fetch promises.
4. Resolve all the promises, add the data to cache.
5. Repeat from step number 3 until we no longer find any promises.
6. We have the HTML markup ready.

This is similar to what `react-apollo` does under the hood. It is quite resource intensive, but until React supports Suspense on SSR, this is what we had to do. 

There is [`react-ssr-prepass`](https://github.com/FormidableLabs/react-ssr-prepass) that solves this by implementing their own partial renderer with custom visitor function. WIth this we could only render once, and wait for data along the way, like Suspense. This is used in [urql](https://github.com/FormidableLabs/urql). 

Unfortunately, I haven't found a way to use that in this project.

**Comment by hindsc52** [comment]
heya, apologies for the delayed reply. Ive been working on integrating this library with [Simorgh](https://github.com/bbc/simorgh) This is the React Isomorphic application used for the BBC World Service.

The progress on 0.15 sounds amazing, The multiple renders on the server were slightly concerning when it comes to performance. I will be sure to try out the updates soon.

Ideally this is something we would love to use and also contribute back too!

**Comment by hindsc52** [comment]
One thing we had noticed recently is that this lib assumes you are fetching JSON (this covers probably 95% of use cases). We have a need to fetch what is ultimately raw HTML from an endpoint do you think it would be possible to remove the the requirement on json data only?

Could this perhaps accept a param to specify if we are fetching JSON or not? https://github.com/jackyef/react-isomorphic-data/blob/master/packages/react-isomorphic-data/src/hooks/utils/useBaseData.ts#L46

happy to raise a PR to support this.

**Comment by hindsc52** [comment]
Hmmm sorry might be missing something here but as react is not yet supported on the server I get the error `Error: ReactDOMServer does not yet support Suspense` How have you got around this in your example?

**Comment by hindsc52** [comment]
I might have done something wring based o0n your examples, I'm going to put together a new PR to try it. Will send it this way if I have any problems.

**Comment by jackyef** [comment]
>We have a need to fetch what is ultimately raw HTML
If I am understanding you correctly, basically the response stored in cache should be the resolved value from `res.text()` instead of `res.json()` right?

I will look into how to go about it. It will probably affect how storing stuffs in cache works.

> Error: ReactDOMServer does not yet support Suspense
Are you, by any chance, using `@loadable/component`? I found similar issues here:
https://github.com/FormidableLabs/urql/issues/488
https://github.com/gregberge/loadable-components/issues/480#issuecomment-573396356

**Comment by chinds185a** [comment]
> If I am understanding you correctly, basically the response stored in cache should be the resolved value from `res.text()` instead of `res.json()` right?

I believe so yes. But it might be worth defaulting to `response.json()` unless something else is specified?

> Are you, by any chance, using @loadable/component? I found similar issues here:
We are yes. I'll look into that issue

**Comment by jackyef** [comment]
Alright! I will try to look into it during this weekend.
https://github.com/jackyef/react-isomorphic-data/issues/50

**Comment by jackyef** [comment]
Hi @chinds185a!

The raw response support is added in https://github.com/jackyef/react-isomorphic-data/pull/51. Just pass `raw: true` in the [`dataOptions`](https://react-isomorphic-data.netlify.com/docs/others/data-options#raw) when using hooks or HOCs.

Released on [0.16.0](https://github.com/jackyef/react-isomorphic-data/releases/tag/react-isomorphic-data%400.16.0)

Upon working on it, I actually found out that there was a very silly bug that is in the code all these times. This release also fixed that bug.

I will close this issue for now because it is starting to deviates from the original topic. Feel free to open a new issue to discuss anything though! Really appreciate your help so far!
